user/src/com/google/gwt/user/server/rpc/RemoteServiceServlet.java

```diff
@@ -366,23 +373,61 @@ public final void processPost(HttpServletRequest request,
     // Let subclasses see the serialized request.
     //
     onBeforeRequestDeserialized(requestPayload);
-
-    // Invoke the core dispatching logic, which returns the serialized
-    // result.
-    //
-    String responsePayload = processCall(requestPayload);
-
-    // Let subclasses see the serialized response.
+    
+    // Create Writer on response, considering GZIP compression if requested
     //
-    onAfterResponseSerialized(responsePayload);
+    final Writer responseWriter = createWriterForResponse(request, response);
+    final boolean onAfterResponseSerializedOverridden = isOnAfterResponseSerializedOverridden();
+    final StringWriter writerForResponseCopy;
+    if (onAfterResponseSerializedOverridden) {
+      writerForResponseCopy = new StringWriter();
+    } else {
+      writerForResponseCopy = null;
+    }
+    final TeeWriter<StringWriter> writer = new TeeWriter<>(responseWriter, writerForResponseCopy);
+    RPC.setResponseWriter(writer);
+    try {
+      // Invoke the core dispatching logic, which returns the serialized
+      // result.
+      //
+      String responsePayload = processCall(requestPayload);
+      
+      if (onAfterResponseSerializedOverridden) {
+        // Let subclasses see the serialized response.
+        // For the old protocol version it was already assembled
+        // and returned by processCall(requestPayload). For protocol
+        // versions starting with AbstractSerializationStream.SERIALIZATION_STREAM_FORWARD_STREAMING_VERSION
+        // null is returned by processCall(requestPayload), and the
+        // payload is obtained from the StringWriter
+        onAfterResponseSerialized(responsePayload != null ? responsePayload : writerForResponseCopy.toString());
+      }
+    } finally {
+      RPC.unsetResponseWriter();
+    }
+  }
 
-    // Write the response.
-    //
-    writeResponse(request, response, responsePayload);
+  private boolean isOnAfterResponseSerializedOverridden() {
+    try {
+      Class<?> c = getClass();
+      boolean foundOverride = false;
+      while (!foundOverride && c != RemoteServiceServlet.class) {
+        try {
+          final Method m = c.getDeclaredMethod("onAfterResponseSerialized", new Class<?>[] { String.class });
+          foundOverride = m != null;
+          // no need to check visibility because an override must be at least protected or public
+        } catch (NoSuchMethodException e) {
+          // ignore; the method wasn't found in the class referenced by c
+        }
+        c = c.getSuperclass();
+      }
+      return foundOverride;
+    } catch (SecurityException e) {
+      throw new RuntimeException("Couldn't find onAfterResponseSerialized method", e);
+    }
   }
 
   /**
-   * This method is called by {@link #processCall(String)} and will throw a
+   * This method is called by {@link #processCall(String, Writer)} and will throw a
```

Comment:
I think you meant to revert this change

Reply:
067d9ed998a1e112d3c021b5994ae28ad9dfcf1b